### PR TITLE
#143

### PR DESCRIPTION
**Pull request by 123joshuawu**
Added `sorts: Option[List[KgNodeSort]]` to `getMatchingNodes`
Added `enum stores.SortDirection`
Added `enum stores.kg.KgNodeField`
Added `case class stores.kg.KgNodeSort(field: KgNodeField, direction: SortDirection)`
Updated graphql `matchingNodes` query

TODO: frontend changes

### CODE REVIEW

File: lib/scala/kg/src/main/scala/models/graphql/AbstractKgGraphQlSchemaDefinition.scala [hunk 1]
```diff
@@ -51,22 +51,37 @@ abstract class AbstractKgGraphQlSchemaDefinition extends BaseGraphQlSchemaDefini
     )
   )
 
+  // Input enum types
+  implicit val KgNodeFieldType = deriveEnumType[KgNodeField.Value](
```
Comment (gordom6) [inline lib/scala/kg/src/main/scala/models/graphql/AbstractKgGraphQlSchemaDefinition.scala:55]: If you use circe + enumeratum this is easier. Here's an example enum definition:

```
import enumeratum.values.{StringCirceEnum, StringEnum, StringEnumEntry}
import sangria.macros.derive.deriveEnumType

sealed abstract class StepOrderFlag(val value: String) extends StringEnumEntry

case object StepOrderFlag extends StringEnum[StepOrderFlag] with StringCirceEnum[StepOrderFlag] {
  case object Causal extends StepOrderFlag("causal")
  case object Optional extends StepOrderFlag("optional")
  case object Precondition extends StepOrderFlag("precondition")
  case object Simultaneous extends StepOrderFlag("simultaneous")
  case object Subevent extends StepOrderFlag("subevent")
  val sangriaType = deriveEnumType[StepOrderFlag]()
  val values = findValues
}
```

You can also use `sealed trait`.

Note that the deriveEnumType is there. You can pull it in as an implicit val = SomeEnum.sangriaType. The derive has to be within the case object in order to pick up all the values. Don't ask me why.

Independently of GraphQL, this style of enums (sealed traits + case objects inside another object) is preferred over Enumeration in Scala.

Comment (gordom6) [inline lib/scala/kg/src/main/scala/models/graphql/AbstractKgGraphQlSchemaDefinition.scala:55]: KgNodeField is probably misnamed. It's not all fields. KgNodeSortableField?

File: lib/scala/kg/src/main/scala/stores/kg/KgNodeField.scala [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  private val ListDelimChar = '|'
+  private val ListDelimString = ListDelimChar.toString
+
+  type KgNodeField = Value
+  val Id = Value("id")
+  val Labels = Value("labels")
+  val Sources = Value("sources")
+  val PageRank = Value("pageRank")
+
+  def getFieldString(node: KgNode, field: KgNodeField): String =
```
Comment (gordom6) [inline lib/scala/kg/src/main/scala/stores/kg/KgNodeField.scala:15]: This is awkward. Put it on KgNode, getFieldString(field: KgNodeField).

Comment (gordom6) [reply lib/scala/kg/src/main/scala/stores/kg/KgNodeField.scala:15]: Your assumption here appears to be incorrect, in any case. Two nodes may have the same set of labels in different order?

File: lib/scala/kg/src/main/scala/stores/kg/KgNodeSort.scala [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import stores.SortDirection
+import stores.SortDirection.SortDirection
+import stores.kg.KgNodeField.KgNodeField
+
+final case class KgNodeSort(field: KgNodeField, direction: SortDirection)
+
+object KgNodeSort {
+  def sort(nodes: List[KgNode], sorts: List[KgNodeSort]): List[KgNode] = {
+    if (sorts.isEmpty) nodes
+    else KgNodeSort.sort(nodes.sortBy(node => KgNodeField.getFieldString(node, sorts.last.field))(if (sorts.last.direction == SortDirection.Ascending) Ordering[String] else Ordering[String].reverse), sorts.init)
```
Comment (gordom6) [inline lib/scala/kg/src/main/scala/stores/kg/KgNodeSort.scala:13]: If this is the only place getFieldString is used, then make it a private method here. You need to treat the fields differently for ordering purposes, anyway. pageRank should be numerically compared, lists should be sorted and then compared, a string should be compared as-is.

KgNodeSort.sort is only used in MemKgQueryStore, anyway, so move this there.

File: lib/scala/kg/src/main/scala/stores/kg/mem/MemKgStore.scala [hunk 4]
```diff
@@ -125,9 +125,9 @@ class MemKgStore extends KgCommandStore with KgQueryStore {
     )
   }
 
-  final override def getMatchingNodes(limit: Int, offset: Int, query: KgNodeQuery): List[KgNode] = {
+  final override def getMatchingNodes(limit: Int, offset: Int, query: KgNodeQuery, sorts: Option[List[KgNodeSort]]): List[KgNode] = {
     val results: PagedResults[SearchResult] = lucene.query().filter(toSearchTerms(query):_*).limit(limit).offset(offset).search()
-    results.results.toList.map(searchResult => nodesById(searchResult(LuceneFields.nodeId)))
+    KgNodeSort.sort(results.results.toList.map(searchResult => nodesById(searchResult(LuceneFields.nodeId))), sorts.getOrElse(List()))
```
Comment (gordom6) [inline lib/scala/kg/src/main/scala/stores/kg/mem/MemKgStore.scala:130]: As I said above, move the .sort method to this class.

File: lib/scala/kg/src/main/scala/stores/kg/neo4j/Neo4jKgQueryStore.scala [hunk 5]
```diff
@@ -159,14 +159,10 @@ final class Neo4jKgQueryStore @Inject()(configuration: Neo4jStoreConfiguration)
...
-    final override def getMatchingNodes(limit: Int, offset: Int, query: KgNodeQuery): List[KgNode] = {
+    final override def getMatchingNodes(limit: Int, offset: Int, query: KgNodeQuery, sorts: Option[List[KgNodeSort]]): List[KgNode] = {
       val cypher = KgNodeQueryCypher(query)
       transaction.run(
-        s"""${cypher}
-           |RETURN ${nodePropertyNamesString}
-           |SKIP ${offset}
-           |LIMIT ${limit}
-           |""".stripMargin,
+        (List(cypher) ++ List(s"RETURN ${nodePropertyNamesString}") ++ (if (sorts.nonEmpty) List(s"ORDER by ${sorts.get.map(sort => s"node.${sort.field.toString} ${sort.direction.toString}").mkString(", ")}") else List()) ++ List(s"SKIP ${offset}") ++ List(s"LIMIT ${limit}")).mkString("\n"),
```
Comment (gordom6) [inline lib/scala/kg/src/main/scala/stores/kg/neo4j/Neo4jKgQueryStore.scala:165]: You took a nice multi-line string and made it unreadable..
It's fine if the Cypher has a blank line in it.
Something like:
${sorts.map(sorts => s"ORDER ...").getOrElse("")}

File: lib/scala/kg/src/main/scala/models/graphql/AbstractKgGraphQlSchemaDefinition.scala (hunk 1 above)
Comment (123joshuawu) [reply lib/scala/kg/src/main/scala/models/graphql/AbstractKgGraphQlSchemaDefinition.scala:55]: Yeah that makes sense, changed to `KgNodeSortableField`

File: lib/scala/kg/src/main/scala/stores/kg/KgNodeField.scala (hunk 2 above)
Comment (123joshuawu) [reply lib/scala/kg/src/main/scala/stores/kg/KgNodeField.scala:15]: Gotcha, added `getFieldString(field: KgNodeSortableField)` to `KgNode`

Comment (123joshuawu) [reply lib/scala/kg/src/main/scala/stores/kg/KgNodeField.scala:15]: And sorted `label` and `sources` before `mkString`. Forgot about the set mechanics

File: lib/scala/kg/src/main/scala/stores/kg/mem/MemKgStore.scala (hunk 4 above)
Comment (123joshuawu) [reply lib/scala/kg/src/main/scala/stores/kg/mem/MemKgStore.scala:130]: I've realized that I'm sorting the results of the query after offset and limit were performed, so I am looking into using the lucene query sort

File: lib/scala/kg/src/main/scala/stores/kg/neo4j/Neo4jKgQueryStore.scala (hunk 5 above)
Comment (123joshuawu) [reply lib/scala/kg/src/main/scala/stores/kg/neo4j/Neo4jKgQueryStore.scala:165]: Yeah that looks a lot better, changed

File: lib/scala/kg/src/main/scala/stores/kg/KgNodeSort.scala (hunk 3 above)
Comment (123joshuawu) [reply lib/scala/kg/src/main/scala/stores/kg/KgNodeSort.scala:13]: Gotcha that makes sense. Am experimenting with Lucene sort to replace this

File: lib/scala/kg/src/main/scala/models/graphql/AbstractKgGraphQlSchemaDefinition.scala (hunk 1 above)
Comment (123joshuawu) [reply lib/scala/kg/src/main/scala/models/graphql/AbstractKgGraphQlSchemaDefinition.scala:55]: Interesting, this is neat, will change

**Pull request merged**